### PR TITLE
Block: new constructors to reduce mutability

### PR DESCRIPTION
**Pull request by msgilligan**
* New constructors and static factory methods
* Make `version` final
* Remove `public` from some @VisibleForTesting setters
* Add @Nullable to `transactions` parameter in public constructor
* Update NetworkParameters classes to use new static factory methods
* Update 2 tests to use new constructors

### CODE REVIEW

**Comment by msgilligan** [comment]
@schildbach Marked as DRAFT because of question of implementation type `ArrayList` or `LinkedList` for `Transaction` list. I think `ArrayList` is probably better.

File: core/src/main/java/org/bitcoinj/core/Block.java [hunk 1]
```diff
@@ -178,13 +179,31 @@ private static List<Transaction> readTransactions(ByteBuffer payload) throws Buf
...
-        time = TimeUtils.currentTime().truncatedTo(ChronoUnit.SECONDS); // convert to Bitcoin time
-        prevBlockHash = Sha256Hash.ZERO_HASH;
+        this(setVersion,
+                TimeUtils.currentTime().truncatedTo(ChronoUnit.SECONDS), // convert to Bitcoin time)
+                0x1d07fff8L,
+                0,
+                Collections.emptyList());
+    }
+
+    // For unit-test genesis blocks
```
Comment (schildbach) [inline core/src/main/java/org/bitcoinj/core/Block.java:193]: Should this be `@VisibleForTesting` then?

**Comment by msgilligan** [comment]
@schildbach rebased and force-pushed.

File: core/src/main/java/org/bitcoinj/core/Block.java (hunk 1 above)
Comment (msgilligan) [reply core/src/main/java/org/bitcoinj/core/Block.java:193]: Yes. I'll update it.

**Comment by schildbach** [comment]
I picked some more changes from this PR.

btw. also see #3079

**Comment by schildbach** [comment]
Merged.

**Pull request closed without merging**